### ISSUE TITLE
[java] New Rule: Disallow overridable methods in object construction

### ISSUE THREAD

**Issue by Fernal73** [pmd/pmd#2988]
**Proposed Rule Name:**
AvoidOverridableMethodsInObjectConstruction

**Proposed Category:** 
Security

**Description:**
https://www.oracle.com/java/technologies/javase/seccodeguide.html#7-4

Guideline 7-4 / OBJECT-4: Prevent constructors from calling methods that can be overridden

"Constructors that call overridable methods give attackers a reference to this (the object being constructed) before the object has been fully initialized. Likewise, clone, readObject, or readObjectNoData methods that call overridable methods may do the same. The readObject methods will usually call java.io.ObjectInputStream.defaultReadObject, which is an overridable method."

https://docs.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca2214

"When a virtual method is called, the actual type that executes the method is not selected until run time. When a constructor calls a virtual method, it's possible that the constructor for the instance that invokes the method has not executed. This could lead to errors or unexpected behavior, if an overridden virtual method relies on initialization and other configuration in the constructor."

https://rules.sonarsource.com/java/RSPEC-1699

"Calling an overridable method from a constructor could result in failures or strange behaviors when instantiating a subclass which overrides the method.

For example:

The subclass class constructor starts by contract by calling the parent class constructor.
The parent class constructor calls the method, which has been overridden in the child class.
If the behavior of the child class method depends on fields that are initialized in the child class constructor, unexpected behavior (like a NullPointerException) can result, because the fields aren't initialized yet."

**Code Sample:** 

```java
public class BadlyConstructedType
{
    protected string initialized = "No";

    public BadlyConstructedType()
    {
        System.out.println("Calling base ctor.");
        // violates rule
        doSomething();
    }
    // This will be overridden in the derived type.
    public  void doSomething()
    {
        System.out.println("Base DoSomething");
    }
}

public class DerivedType extends BadlyConstructedType
{
    public DerivedType()
    {
        System.out.printlln("Calling derived ctor.");
        initialized = "Yes";
    }


@Override
    public void doSomething()
    {
        System.out.printf("Derived doSomething is called - initialized ? %s%n", initialized);
    }
}

public class TestBadlyConstructedType
{
    public static void main(String... args)
    {
        DerivedType derivedInstance = new DerivedType();
    }
}
```

**Possible Properties:**

checkCloneMethod (Boolean)
checkReadObjectMethods (Boolean)

*   Should this rule be customizable via properties?
Yes. See above.

**Comment by oowekyala** [comment]
[ConstructorCallsOverridableMethod](https://pmd.github.io/latest/pmd_rules_java_errorprone.html#constructorcallsoverridablemethod)

**Comment by Fernal73** [comment]
@oowekyala I seem to have missed that. Closing issue, will reopen it as an enhancement in separate issue.
